File: app/src/main/java/org/nevack/unitconverter/ConverterDisplayView.java
package org.nevack.unitconverter;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.app.Activity;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.graphics.Color;
import android.graphics.Rect;
import android.os.Build;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v4.content.ContextCompat;
import android.text.Editable;
import android.text.Html;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewAnimationUtils;
import android.view.ViewGroupOverlay;
import android.view.animation.AccelerateDecelerateInterpolator;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.SpinnerAdapter;
import android.widget.TextView;

import org.nevack.unitconverter.model.Unit;

import java.util.List;

import static android.content.Context.CLIPBOARD_SERVICE;
import static org.nevack.unitconverter.R.layout.display;


public class ConverterDisplayView extends LinearLayout {

    private static final String COPY_RESULT = "converter_result";

    private Spinner sourceSpinner;
    private EditText sourceEditText;
    private TextView sourceTextView;
    private EditText resultEditText;
    private TextView resultTextView;
    private Spinner resultSpinner;
    private FloatingActionButton fab;

    private List<Unit> units;

    interface DisplayCallback {
        String convert(String value, int from, int to);
    }

    private DisplayCallback callback;

    public ConverterDisplayView(final Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);

        callback = (DisplayCallback) context;

        setOrientation(VERTICAL);

        inflate(context, display, this);

        fab = (FloatingActionButton) findViewById(R.id.fab);

        sourceTextView = (TextView) findViewById(R.id.sourcevaluesymbol);
        resultTextView = (TextView) findViewById(R.id.resultvaluesymbol);

        sourceSpinner = (Spinner) findViewById(R.id.sourcespinner);
        sourceSpinner.setOnItemSelectedListener(new SpinnerListener(sourceTextView));

        resultSpinner = (Spinner) findViewById(R.id.resultspinner);
        resultSpinner.setOnItemSelectedListener(new SpinnerListener(resultTextView));

        sourceEditText = (EditText) findViewById(R.id.sourcevalue);
        sourceEditText.addTextChangedListener(new TextWatcher() {
            @Override public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override public void onTextChanged(CharSequence s, int start, int before, int count) {}

            @Override
            public void afterTextChanged(Editable s) {
                if (sourceEditText.getText().toString().equals(".")) {
                    sourceEditText.setText("0.");
                    sourceEditText.setSelection(sourceEditText.getText().toString().length());
                }

                if (sourceEditText.getText().toString().equals("-.")) {
                    sourceEditText.setText("-0.");
                    sourceEditText.setSelection(sourceEditText.getText().toString().length());
                }

                if (!sourceEditText.getText().toString().equals("-")) convert();
            }
        });
        resultEditText = (EditText) findViewById(R.id.resultvalue);

        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String temp = sourceEditText.getText().toString();
                sourceEditText.setText(resultEditText.getText().toString());
                resultEditText.setText(temp);
                int position = sourceSpinner.getSelectedItemPosition();
                sourceSpinner.setSelection(resultSpinner.getSelectedItemPosition());
                resultSpinner.setSelection(position);
                convert();
            }
        });

        sourceTextView.setOnClickListener(new View.OnClickListener() {
            @Override public void onClick(View v) { sourceSpinner.performClick(); }
        });

        resultTextView.setOnClickListener(new View.OnClickListener() {
            @Override public void onClick(View v) { resultSpinner.performClick(); }
        });
    }

    private void setSpinnerAdapter(SpinnerAdapter adapter) {
        sourceSpinner.setAdapter(adapter);
        resultSpinner.setAdapter(adapter);
    }

    public void setUnits(List<Unit> units) {
        this.units = units;
        ArrayAdapter<String> adapter =
                new ArrayAdapter<>(getContext(), android.R.layout.simple_list_item_1);
        for (Unit unit : this.units) {
            adapter.add(unit.getName());
        }

        setSpinnerAdapter(adapter);
    }

    public void erase() {
        ViewGroupOverlay groupOverlay = null;
        View revealView = new View(getContext());
        Animator revealAnimator = null;

        final Rect displayRect = new Rect();
        this.getGlobalVisibleRect(displayRect);

        // Make reveal cover the display and status bar.
        revealView.setBottom(displayRect.bottom);
        revealView.setLeft(displayRect.left);
        revealView.setRight(displayRect.right);
        revealView.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.colorAccent));

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
            groupOverlay = (ViewGroupOverlay) ((Activity) getContext()).getWindow().getDecorView().getOverlay();
            groupOverlay.add(revealView);
        }
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
            final int[] clearLocation = new int[2];
            this.getLocationInWindow(clearLocation);
            clearLocation[0] += this.getWidth();
            clearLocation[1] += this.getHeight();

            final int revealCenterX = clearLocation[0] - revealView.getLeft();
            final int revealCenterY = clearLocation[1] - revealView.getTop();

            final double x1_2 = Math.pow(revealView.getLeft() - revealCenterX, 2);
            final double x2_2 = Math.pow(revealView.getRight() - revealCenterX, 2);
            final double y_2 = Math.pow(revealView.getTop() - revealCenterY, 2);
            final float revealRadius = (float) Math.max(Math.sqrt(x1_2 + y_2), Math.sqrt(x2_2 + y_2));

            revealAnimator = ViewAnimationUtils.createCircularReveal(revealView,
                    revealCenterX, revealCenterY, 0.0f, revealRadius);

            revealAnimator.setDuration(
                    getResources().getInteger(android.R.integer.config_longAnimTime));
        }

        final Animator alphaAnimator = ObjectAnimator.ofFloat(revealView, View.ALPHA, 0.0f);
        alphaAnimator.setDuration(
                getResources().getInteger(android.R.integer.config_mediumAnimTime));
        alphaAnimator.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationStart(Animator animation) {
                resultEditText.getEditableText().clear();
                sourceEditText.getEditableText().clear();
            }
        });

        final AnimatorSet animatorSet = new AnimatorSet();

        //Play Reveal Animation if Lollipop or higher, and only alpha animation for others
        if (revealAnimator != null) {
            animatorSet.play(revealAnimator).before(alphaAnimator);
        } else animatorSet.play(alphaAnimator);

        animatorSet.setInterpolator(new AccelerateDecelerateInterpolator());

        final ViewGroupOverlay finalGroupOverlay = groupOverlay;
        final View finalRevealView = revealView;

        animatorSet.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animator) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
                    if (finalGroupOverlay != null) {
                        finalGroupOverlay.remove(finalRevealView);
                    }
                }
            }
        });

        animatorSet.start();
    }

    private void convert() {
        resultEditText.setText(
                callback.convert(sourceEditText.getText().toString(),
                        sourceSpinner.getSelectedItemPosition(),
                        resultSpinner.getSelectedItemPosition())
        );
    }

    public void copyResultToClipboard(boolean withUnitSymbol) {
        String textToCopy = "";
        if (!resultEditText.getText().toString().equals("")) {
            textToCopy = resultEditText.getText().toString();
        }
        if (withUnitSymbol)
        {
            textToCopy += resultTextView.getText().toString();
        }
        ClipboardManager clipboard = (ClipboardManager)
                getContext().getSystemService(CLIPBOARD_SERVICE);
        clipboard.setPrimaryClip(ClipData.newPlainText(COPY_RESULT, textToCopy));
        Snackbar.make(this, R.string.copy_result_toast, Snackbar.LENGTH_SHORT).show();
    }

    public void setupWithKeypad(KeypadView view) {
        view.setEditText(sourceEditText);

        view.setCopyListener(
                new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        copyResultToClipboard(false);
                    }
                },
                new View.OnLongClickListener() {
                    @Override
                    public boolean onLongClick(View v) {
                        copyResultToClipboard(true);
                        return true;
                    }
                }
        );

        view.setBackspaceLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                erase();
                return true;
            }
        });
    }

    private class SpinnerListener implements Spinner.OnItemSelectedListener {

        private TextView textView;

        SpinnerListener(TextView textView) {
            this.textView = textView;
        }

        @Override
        public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
            ((TextView) parent.getChildAt(0)).setTextColor(Color.WHITE);
            textView.setText(Html.fromHtml(units.get(parent.getSelectedItemPosition()).getUnitSymbol()));
            convert();
        }

        @Override public void onNothingSelected(AdapterView<?> parent) { convert(); }
    }
}

File: app/src/main/kotlin/dev/nevack/unitconverter/converter/ConverterDisplayView.kt
package dev.nevack.unitconverter.converter

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewAnimationUtils
import android.view.ViewGroup
import android.view.ViewGroupOverlay
import android.view.animation.AccelerateDecelerateInterpolator
import android.widget.ArrayAdapter
import android.widget.LinearLayout
import androidx.core.text.parseAsHtml
import androidx.core.widget.doAfterTextChanged
import com.google.android.material.textfield.TextInputLayout
import dev.nevack.unitconverter.R
import dev.nevack.unitconverter.databinding.DisplayBinding
import dev.nevack.unitconverter.model.ConversionUnit
import kotlin.math.pow
import kotlin.math.sqrt

internal class ConverterDisplayView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0,
    defStyleRes: Int = 0,
) : LinearLayout(context, attrs, defStyleAttr, defStyleRes) {
    private val binding: DisplayBinding
    private var conversionUnits: MutableList<ConversionUnit> = mutableListOf()

    private var sourceIndex = 0
        set(value) {
            field = value
            updateSuffixes()
        }
    private var resultIndex = 0
        set(value) {
            field = value
            updateSuffixes()
        }

    init {
        orientation = VERTICAL
        binding = DisplayBinding.inflate(LayoutInflater.from(context), this).apply {
            fab.setOnClickListener {
                convertData = convertData.swap()
            }
            sourceValue.showSoftInputOnFocus = false
            resultValue.showSoftInputOnFocus = false
        }
    }

    fun onTextChanged(watcher: (text: String) -> Unit) {
        binding.sourceValue.doAfterTextChanged {
            it?.run { watcher(toString()) }
        }
    }

    fun showResult(result: String) {
        binding.resultValue.setText(result)
    }

    var convertData: ConvertData
        get() = with(binding) {
            ConvertData(
                sourceValue.text.toString(),
                resultValue.text.toString(),
                sourceIndex,
                resultIndex
            )
        }
        set(convertData) = with(binding) {
            sourceValue.setText(convertData.value)
            resultValue.setText(convertData.result)
            sourceIndex = convertData.from
            resultIndex = convertData.to
            sourceSpinner.setText(conversionUnits[sourceIndex].name, false)
            resultSpinner.setText(conversionUnits[resultIndex].name, false)
        }

    fun getCopyResult(withUnitSymbols: Boolean): String {
        return binding.resultValue.text.toString() +
            if (withUnitSymbols) binding.sourceValueContainer.suffixText else ""
    }

    private fun setSpinnerAdapter(adapter: ArrayAdapter<String>, isDifferent: Boolean) {
        binding.sourceSpinner.setAdapter(adapter)
        binding.resultSpinner.setAdapter(adapter)
        if (!adapter.isEmpty) {
            sourceIndex = 0
            resultIndex = if (isDifferent) 1 else 0
            binding.sourceSpinner.setText(adapter.getItem(sourceIndex), false)
            binding.resultSpinner.setText(adapter.getItem(resultIndex), false)
        } else {
            binding.sourceSpinner.isEnabled = false
            binding.resultSpinner.isEnabled = false
        }
    }

    fun setUnits(units: List<ConversionUnit>) {
        conversionUnits.clear()
        conversionUnits.addAll(units)
        val adapter = ArrayAdapter(
            context,
            android.R.layout.simple_list_item_1,
            conversionUnits.map { it.name }
        )
        setSpinnerAdapter(adapter, units.size > 1)
    }

    fun erase() {
        val revealView = View(context)
        // Make reveal cover the display and status bar.
        revealView.top = -y.toInt()
        revealView.bottom = bottom
        revealView.left = left
        revealView.right = right
        revealView.setBackgroundColor(context.getColor(R.color.colorSecondary))
        val groupOverlay = (parent as ViewGroup).overlay as ViewGroupOverlay
        groupOverlay.add(revealView)
        val revealCenterX = width
        val revealCenterY = height + y.toInt()
        val revealRadius = sqrt(revealCenterX.toFloat().pow(2) + revealCenterY.toFloat().pow(2))
        val revealAnimator = ViewAnimationUtils.createCircularReveal(
            revealView,
            revealCenterX, revealCenterY, 0.0f, revealRadius
        )
        revealAnimator.duration =
            resources.getInteger(android.R.integer.config_longAnimTime).toLong()
        val alphaAnimator: Animator = ObjectAnimator.ofFloat(revealView, ALPHA, 0.0f)
        alphaAnimator.duration =
            resources.getInteger(android.R.integer.config_mediumAnimTime).toLong()
        alphaAnimator.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationStart(animation: Animator) {
                clear()
            }
        })
        val animatorSet = AnimatorSet()

        animatorSet.play(revealAnimator).before(alphaAnimator)
        animatorSet.interpolator = AccelerateDecelerateInterpolator()
        animatorSet.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animator: Animator) {
                groupOverlay.remove(revealView)
            }
        })
        animatorSet.start()
    }

    fun clear() {
        binding.resultValue.editableText.clear()
        binding.sourceValue.editableText.clear()
    }

    fun appendText(text: CharSequence) {
        binding.sourceValue.append(text)
    }

    fun removeLastDigit() {
        val textString = binding.sourceValue.text.toString()
        if (textString.isNotEmpty()) {
            binding.sourceValue.setText(textString.substring(0, textString.length - 1))
            binding.sourceValue.selectAll()
        }
    }

    fun setSpinnersCallback(callback: () -> Unit) {
        binding.sourceSpinner.setOnItemClickListener { _, _, position, _ ->
            sourceIndex = position
            callback()
        }
        binding.resultSpinner.setOnItemClickListener { _, _, position, _ ->
            resultIndex = position
            callback()
        }
    }

    private fun updateSuffixes() {
        binding.sourceValueContainer.setHtmlSuffixText(conversionUnits[sourceIndex].unitSymbol)
        binding.resultValueContainer.setHtmlSuffixText(conversionUnits[resultIndex].unitSymbol)
    }

    private fun TextInputLayout.setHtmlSuffixText(text: String) {
        suffixText = text
        suffixTextView.text = text.parseAsHtml()
    }
}
